File: June-2023/C++/NumberOfProvinces.cxx
// 547. Number of Provinces
// simple dfs
class Solution {
    int count = 0;

    void dfs(unordered_map<int, list<int>>& map, unordered_set<int>& visited, int u) {
        if (visited.count(u)) return;
        for (int v : map[u]) {
            if (!visited.count(v)) {
                visited.insert(u);
                dfs(map, visited, v);
            }
        }
    }

public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        unordered_map<int, list<int>> map;
        for (int i = 0; i < isConnected.size(); i++) {
            for (int j = 0; j < isConnected[0].size(); j++) {
                if (isConnected[i][j] == 1) {
                    if (map.find(i) == map.end()) {
                        map[i] = list<int>();
                    }
                    map[i].push_back(j);
                }
            }
        }
        unordered_set<int> visited;
        for (int i = 0; i < isConnected.size(); i++) {
            if (!visited.count(i)) {
                dfs(map, visited, i);
                count++;
            }
        }
        return count;
    }
};

File: June-2023/Java/NumberOfProvinces.java
// 547. Number of Provinces
// simple dfs
class Solution {
	int count = 0;
	public void dfs(HashMap<Integer, List<Integer>> map, Set<Integer> visited, int u) {
		if (visited.contains(u)) return;
		for (int v : map.getOrDefault(u, new LinkedList<>())) {
			if (!visited.contains(v)) {
				visited.add(u);
				dfs(map, visited, v);
			}
		}
	}
	public int findCircleNum(int[][] isConnected) {
		HashMap<Integer, List<Integer>> map = new HashMap<>();
		for (int i = 0; i < isConnected.length; i++) {
			for (int j = 0; j < isConnected[0].length; j++) {
				if (isConnected[i][j] == 1) {
					if (!map.containsKey(i)) {
						map.put(i, new LinkedList<>());
					}
					map.get(i).add(j);
				}
			}
		}
		Set<Integer> visited = new HashSet<>();
		for (int i = 0; i < isConnected.length; i++) {
			if (!visited.contains(i)) {
				dfs(map, visited, i);
				count++;
			}
		}
		return count;
	}
}
